crates/exex/exex/src/backfill.rs

```diff
@@ -0,0 +1,355 @@
+use reth_db_api::database::Database;
+use reth_evm::execute::{BatchExecutor, BlockExecutionError, BlockExecutorProvider};
+use reth_node_api::FullNodeComponents;
+use reth_primitives::{Block, BlockNumber};
+use reth_provider::{
+    Chain, FullProvider, HistoricalStateProviderRef, ProviderError, TransactionVariant,
+};
+use reth_prune_types::PruneModes;
+use reth_revm::database::StateProviderDatabase;
+use reth_stages_api::{format_gas_throughput, ExecutionStageThresholds};
+use reth_tracing::tracing::{debug, trace};
+use std::{
+    marker::PhantomData,
+    ops::RangeInclusive,
+    time::{Duration, Instant},
+};
+
+/// Factory for creating new backfill jobs.
+#[derive(Debug, Clone)]
+pub struct BackfillJobFactory<E, P> {
+    executor: E,
+    provider: P,
+    prune_modes: PruneModes,
+    thresholds: ExecutionStageThresholds,
+}
+
+impl<E: Clone, P: Clone> BackfillJobFactory<E, P> {
+    /// Creates a new [`BackfillJobFactory`].
+    pub const fn new(
+        executor: E,
+        provider: P,
+        prune_modes: PruneModes,
+        thresholds: ExecutionStageThresholds,
+    ) -> Self {
+        Self { executor, provider, prune_modes, thresholds }
+    }
+
+    /// Creates a new backfill job for the given range.
+    pub fn backfill<DB>(&self, range: RangeInclusive<BlockNumber>) -> BackfillJob<E, DB, P> {
+        BackfillJob {
+            executor: self.executor.clone(),
+            provider: self.provider.clone(),
+            prune_modes: self.prune_modes.clone(),
+            range,
+            thresholds: self.thresholds.clone(),
+            _db: PhantomData,
+        }
+    }
+}
+
+impl BackfillJobFactory<(), ()> {
+    /// Creates a new [`BackfillJobFactory`] from [`FullNodeComponents`].
+    pub fn new_from_components<Node: FullNodeComponents>(
+        components: Node,
+        prune_modes: PruneModes,
+        thresholds: ExecutionStageThresholds,
+    ) -> BackfillJobFactory<Node::Executor, Node::Provider> {
+        BackfillJobFactory::<_, _>::new(
+            components.block_executor().clone(),
+            components.provider().clone(),
+            prune_modes,
+            thresholds,
+        )
+    }
+}
+
+/// Backfill job started for a specific range.
+///
+/// It implements [`Iterator`] that executes blocks in batches according to the provided thresholds
+/// and yields [`Chain`]
+#[derive(Debug)]
+pub struct BackfillJob<E, DB, P> {
+    executor: E,
+    provider: P,
+    prune_modes: PruneModes,
+    range: RangeInclusive<BlockNumber>,
+    thresholds: ExecutionStageThresholds,
+    _db: PhantomData<DB>,
+}
+
+impl<E, DB, P> Iterator for BackfillJob<E, DB, P>
+where
+    E: BlockExecutorProvider,
+    DB: Database,
+    P: FullProvider<DB>,
```

Comment:
this should only only the traits we really need here, because I think we don't need the `DB` type?

Reply:
yeah, as you can see, we need it anyway for `database_provider_ro`, so maybe just the `FullProvider` is fine? or do you think it makes sense to restrict it only to needed traits?